demos/metrics_demo.cc

```diff
@@ -0,0 +1,137 @@
+            // A function that can return any integer
+            sm::make_counter("function", [this]() {applog.info("triggered"); return ++_triggered;}, sm::description("A function which returns a value.")),
+            // A gauge is a counter that can go up and down, can be used for rates
+            sm::make_gauge("gauge", []() {return 3;}, sm::description("random number"))
+        });
+    }
+
+    void increase(uint64_t val = 1) { _counter += val; }
+};
+
+class PrometheusExample {
+private:
+    httpd::http_server_control _server;
+    bool _was_started;
+public:
+    PrometheusExample(sstring address, uint16_t port, sstring prefix) {
+        prometheus::config pctx;
+        net::inet_address prom_addr(address);
+        pctx.metric_help = "seastar::httpd server statistics";
+
+        std::cout << "starting prometheus API server" << std::endl;
+        _server.start("prometheus").get();
+        prometheus::start(_server, pctx).get();
+        _was_started = true;
+
+        _server.listen(socket_address{prom_addr, port}).handle_exception([prom_addr, port] (auto ep) {
+            std::cerr << seastar::format("Could not start Prometheus API server on {}:{}: {}\n", prom_addr, port, ep);
+            return make_exception_future<>(ep);
+        }).get();
+    }
+
+    ~PrometheusExample() {
+        if (_was_started) {
+            std::cout << "Stoppping Prometheus server" << std::endl;  // This can throw, but won't.
+            _server.stop().get();
+        }
+    }
+};
+
+/**
+ * In order to access the metrics we need to use one of the builtin mechanisms.
+ * For the sake of this example we are using prometheus.
+ * To view the specific metrics of the demo use something in the lines of
+ *  curl http://localhost:9180/metrics | grep "metrics_demo"
+ * 
+ */
+int main(int argc, char** argv) {
+    seastar::app_template app;
+    app.add_options()("prometheus_port", bpo::value<uint16_t>()->default_value(9180), "Prometheus port. Set to zero in order to disable.");
+    app.add_options()("prometheus_address", bpo::value<sstring>()->default_value("0.0.0.0"), "Prometheus address");
+    app.add_options()("prometheus_prefix", bpo::value<sstring>()->default_value("seastar_httpd"), "Prometheus metrics prefix");
+
+    return app.run(argc, argv, [&] {
+        return seastar::async([&] {
+            auto&& config = app.configuration();
+            uint16_t pport = config["prometheus_port"].as<uint16_t>();
+            sstring address = config["prometheus_address"].as<sstring>();
+            sstring prefix = config["prometheus_prefix"].as<sstring>();
+            PrometheusExample server(address, pport, prefix);
+
+            applog.info("Starting with metrics 1");
+            {
+                MetricsExample metrics(1);
+                for (int i = 0; i < 10; i++) {
+                    sleep(1s).get();
+                    metrics.increase(5);
+                }
+            }
```

Comment:
I don't quite get what the above block is for. It will just wipe everything without a trace in 10 seconds

Reply:
The point is to demonstrate that when the containing object gets out of scope then the metrics are no longer reported.
While iterating it on it I simply used `curl http://localhost:9180/metrics`, to see the data live.
My thoughts were on to put an emphasis that its most likely preferable to keep the object alive as long as some monitoring system cares about it.

Do you think this point is clear without this part? Any alternative approach to explain this? or eliminate it all together?

Reply:
I would eliminate it altogether and made the demo just start the server "forever" and printing a message to the user saying something like

```
Now try `curl http://127.0.0.1:123/metrics` to see how metrics are exported
```